azkaban-common/src/main/java/azkaban/executor/RunningExecutionsUpdaterThread.java

```diff
@@ -0,0 +1,293 @@
+package azkaban.executor;
+
+import azkaban.alert.Alerter;
+import azkaban.metrics.CommonMetrics;
+import azkaban.utils.JSONUtils;
+import azkaban.utils.Pair;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import javax.inject.Inject;
+import org.apache.log4j.Logger;
+
+public class RunningExecutionsUpdaterThread extends Thread {
+
+  private static final Logger logger = Logger.getLogger(RunningExecutionsUpdaterThread.class);
+
+  private final int waitTimeIdleMs = 2000;
+  private final int waitTimeMs = 500;
+  // When we have an http error, for that flow, we'll check every 10 secs, 360
+  // times (3600 seconds = 1 hour) before we send an email about unresponsive executor.
+  private final int numErrorsBetweenUnresponsiveEmail = 360;
+  // First email is sent after 1 minute of unresponsiveness
+  private final int numErrorsBeforeUnresponsiveEmail = 6;
+  private final long errorThreshold = 10000;
+  private final ExecutorManagerUpdaterStage updaterStage;
+  private final AlerterHolder alerterHolder;
+  private final CommonMetrics commonMetrics;
+  private final ExecutorApiGateway apiGateway;
+  private final RunningExecutions runningExecutions;
+  private final ExecutionFinalizer executionFinalizer;
+  private boolean shutdown = false;
+  private long lastThreadCheckTime = -1;
+
+  @Inject
+  public RunningExecutionsUpdaterThread(final ExecutorManagerUpdaterStage updaterStage,
+      final AlerterHolder alerterHolder, final CommonMetrics commonMetrics,
+      final ExecutorApiGateway apiGateway, final RunningExecutions runningExecutions,
+      final ExecutionFinalizer executionFinalizer) {
+    this.updaterStage = updaterStage;
+    this.alerterHolder = alerterHolder;
+    this.commonMetrics = commonMetrics;
+    this.apiGateway = apiGateway;
+    this.runningExecutions = runningExecutions;
+    this.executionFinalizer = executionFinalizer;
+    this.setName("ExecutorManagerUpdaterThread");
+  }
+
+  void shutdown() {
+    this.shutdown = true;
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public void run() {
+    while (!this.shutdown) {
+      try {
+        this.lastThreadCheckTime = System.currentTimeMillis();
+        this.updaterStage.set("Starting update all flows.");
+        // TODO refresh active executors
+        // TODO BUT this needs to know the up-to-date list of active AND inactive executors. OR:
+        // TODO OR then this would need to call getFlowToExecutorMap() again (which returns all
+        // executors regardless of active=true|false - or null optional if it doesn't exist)
+        // ..so would that make "refresh active executors" just a side-effect?
```

Comment:
I don't quite follow the TODO here. Do you intend to load all executors or just active executors?

Reply:
The thing is I'm not so sure what would be the best way, but I'll think about it. The main point for having this TODO here is to remind that checking for _active_ executors only is not enough.

What I'm trying to say there is that one option would be call `getFlowToExecutorMap()` if executor can't be reached. Because the result of that can be used to check if the executor doesn't exist any more.

On the other hand reloading _active_ executors won't be enough for this check, because active executors don't include inactive executors. This class needs to know if the executor has been removed entirely.

Outside of this class there's a need for updating _active_ executors: for the dispatching side to take new executors into use without manually asking to refresh via API or restarting azkaban-web.

I'm wondering what would be the cleanest way to achieve both things. The ways mentioned above are nice in a way that they're pretty much additive in nature. Other approaches might involve more refactoring, but result in a cleaner outcome. For example, maybe it would make sense to load both inactive and active executors at the same time, and then filter them to only use the active ones in dispatching side.

Reply:
> What I'm trying to say there is that one option would be call getFlowToExecutorMap() if executor can't be reached.

Currently, `getFlowToExecutorMap()` will only check the runningExecutions cache to get the result. However in the previous case we discussed where an executor was shut down first and the runningExecutions cache didn't get a chance to update, this would not solve the problem. 
How about checking the executor info from DB directly if that executor cannot be reached? Then you wouldn't need to reload active executors.

> Outside of this class there's a need for updating active executors: for the dispatching side to take new executors into use without manually asking to refresh via API or restarting azkaban-web.

This would be an improvement but not necessarily the fix for the current issue. Also in our case, updating new executors is only required after new executor deployment, so I don't see much efficiency in updating frequently in the updater thread. 
Correct me if I understand it wrong.

Reply:
> Currently, getFlowToExecutorMap() will only check the runningExecutions cache to get the result.

I'm sorry, I somehow confused it with `ExecutorManager#loadRunningExecutions()`, which is only called when the ExecutorManager is created.

> How about checking the executor info from DB directly if that executor cannot be reached? Then you wouldn't need to reload active executors.

Yeah, why not. I was just thinking of some way to possibly combine shared needs. At least I would implement that check so that I would call the existing `executorLoader.fetchActiveExecutors()` and check if the result includes the executor or not, instead of adding new DB-level code to check for the existence of a specific executor id.

> Then you wouldn't need to reload active executors.
> This would be an improvement but not necessarily the fix for the current issue.

Yes, exactly. I also came to the conclusion that these are separate things, and it may not even make sense to combine them.

> Also in our case, updating new executors is only required after new executor deployment, so I don't see much efficiency in updating frequently in the updater thread.

That is true, but then it would require user to call the refresh executors API, which complicates the operating side. My point was that listing executors periodically is not expensive at all because it's such a small table, and doing that often enough would make adding and removing executors more convenient.